R/utils.R

```diff
@@ -327,4 +361,79 @@ valid_seq <- function(params) {
   } else {
     return(FALSE)
   }
-}
\ No newline at end of file
+}
+
+resolve_prefix <- function(prefix, environment_var) {
+  if (!(is.null(prefix))) {
+    assertthat::assert_that(is.character(prefix))
+    
+    return(prefix)
+  } else {
+    prefix_env <- Sys.getenv(environment_var)
+    if (prefix_env != "") {
+      return(prefix_env)
+    } else {
+      return("/")
+    }
+  }
+}
+
+# The function below requires a dependency path, package information
+# retrieved from a request URL, as well as a list of dependencies
+# (currently in htmltools htmlDependency format). get_package_mapping
+# optionally returns an R package name (if the file is contained
+# inside an R package), or NULL if the dependency is not found,
+# and a (local) path to the dependency.
+# 
+# script_name is e.g. "dash_core_components.min.js"
+# url_package is e.g. "dash_core_components"
+# dependencies = list of htmlDependency objects
+# this function returns a list with two elements:
+#   rpkg_name = character string supplying the name of the R package
+#   rpkg_path = character string providing the path to the dependency
+get_package_mapping <- function(script_name, url_package, dependencies) {
+  # TODO: improve validation of dependency inputs, particularly
+  #       to avoid duplicating dependencies in the package_map
+  package_map <- vapply(unique(dependencies), function(x) {
+    if (x$name %in% c('react', 'react-dom')) {
+      x$name <- 'dash-renderer'
+    }
+    dep_path <- paste(x$src$file, 
+                      x$script,
+                      sep = "/")
+  
+    # remove n>1 slashes and replace with / if present
```

Comment:
could you prefix this comment with something like "htmltools sometimes includes multiple `//` - replace these with `/`". 
That way, the reader knows that `//` is coming from htmltools and not necessarily the dash-renderer API. In the future if you refactor out `htmltools`, then the reader will know that this `//` won't need to be handled because `htmltools` was the source of the issue.

Reply:
I've started the process of :hocho: `htmltools`, managed to remove nearly all references to this package, but not quite there yet. I opened an issue earlier this week which attempts to enumerate all the spots in the code and comments that are involved:

https://github.com/plotly/dashR/issues/54

Reply:
Yeah - My suggestion here is mostly just to get in the habit of writing these types of commits - if there is something in the code that looks odd, it's good to get in the habit of explaining _why_ we have to do stuff like this and where the problem originated.

In this case, everyone involved in this PR is now aware of the issue and you will be the one refactoring it shortly. However, imagine the case where this refactor might happen 6 months from now _by someone else_. If they didn't know that this `//` was originating from `htmltools`, then they might not feel comfortable removing this block of code. Perhaps they might think that `dash-renderer` (a 3rd party piece of code) _sometimes_ introduces a `//` in certain scenarios, scenarios that might not necessarily be tested in integration tests.

So, it's not strictly necessary in this scenario where we plan on fixing this up in a couple of weeks, but it's a good habit to get into and a healthy way to start thinking about comments

Reply:
For sure, decisions like these seem apparent now that I'm staring at the same code over and over again, but for anyone else (or me even a year from now), statements like this are hard to comprehend in total isolation.

I'll be a bit more diligent about making these sorts of commits, it benefits everyone from the author, to the reviewer, and even savvy users who decide to take a look under the hood.

Reply:
fixed in b77c94d57625d31e9914511943ea730e6fc4e86d